src/analysis/monotone-analyzer.h

```diff
@@ -0,0 +1,59 @@
+#ifndef wasm_analysis_monotone_analyzer_h
+#define wasm_analysis_monotone_analyzer_h
+
+#include <iostream>
+#include <queue>
+#include <vector>
+
+#include "cfg.h"
+#include "lattice.h"
+#include "wasm-traversal.h"
+
+namespace wasm::analysis {
+
+template<size_t N> struct MonotoneCFGAnalyzer;
+
+template<size_t N>
+struct BlockState : public UnifiedExpressionVisitor<BlockState<N>> {
```

Comment:
`BlockState` currently bundles multiple concepts for a single block together:
  - The dependent blocks
  - The transfer function
  - The states associated with a block

This is a sort of object-oriented approach where all the state for a single object (i.e block) is put together and represented by a single class, but the downside is that putting all the state together makes it difficult to expose and configure different pieces separately. For example, the dependent blocks and transfer function will be specific to each analysis, so they should be provided by the framework user, but the user shouldn't also need to manage the intermediate lattice states for each block, since that can be the same for every analysis. Having the analysis framework separately manage the dependency graph, transfer function, and intermediate states will make exposing the right interface to the user much simpler.

It's fine if you want to land this with the current design and iterate afterward, though.

Reply:
This will be left to follow-up PRs where I will try to separate responsibilities between different classes and extract common elements.